compiler/grainformat/reformat.re

```diff
@@ -0,0 +1,2753 @@
+          false,
+        );
+      } else {
+        (
+          Doc.concat([
+            print_ident(location.txt),
+            if (List.length(patterns) > 0) {
+              add_parens(
+                Doc.join(
+                  Doc.comma,
+                  List.map(
+                    pat => print_pattern(~pat, ~parent_loc, ~original_source),
+                    patterns,
+                  ),
+                ),
+              );
+            } else {
+              Doc.nil;
+            },
+          ]),
+          false,
+        );
+      };
+
+    | PPatOr(pattern1, pattern2) => (
+        Doc.text("/* Formatter PPatOr error */"),
+        false,
+      )
+    | PPatAlias(pattern, loc) => (
+        Doc.text("/* Formatter PPatAlias error */"),
+        false,
+      )
+    };
+
+  let (pattern, parens) = printed_pattern;
+
+  let with_leading =
+    if (leading_comment_docs == Doc.nil) {
+      [pattern];
+    } else {
+      [leading_comment_docs, Doc.space, pattern];
+    };
+  let with_trailing =
+    if (trailing_comment_docs == Doc.nil) {
+      with_leading;
+    } else {
+      List.append(with_leading, [trailing_comment_docs]);
+    };
+
+  let clean_pattern =
+    if (List.length(with_trailing) == 1) {
+      List.hd(with_trailing);
+    } else {
+      Doc.concat(with_trailing);
+    };
+
+  if (parens) {
+    Doc.concat([Doc.lparen, clean_pattern, Doc.rparen]);
+  } else {
+    clean_pattern;
+  };
+}
+and print_constant = (c: Parsetree.constant) => {
+  let print_c =
+    switch (c) {
+    | PConstNumber(PConstNumberInt(i)) => Printf.sprintf("%s", i)
+    | PConstNumber(PConstNumberFloat(f)) => Printf.sprintf("%s", f)
+    | PConstNumber(PConstNumberRational(n, d)) =>
+      Printf.sprintf("%s/%s", n, d)
+    | PConstBytes(b) => Printf.sprintf("%s", b)
+    | PConstChar(c) => Printf.sprintf("'%s'", String.escaped(c))
+    | PConstFloat64(f) => Printf.sprintf("%sd", f)
+    | PConstFloat32(f) => Printf.sprintf("%sf", f)
+    | PConstInt32(i) => Printf.sprintf("%sl", i)
+    | PConstInt64(i) => Printf.sprintf("%sL", i)
+    | PConstWasmI32(i) => Printf.sprintf("%sn", i)
+    | PConstWasmI64(i) => Printf.sprintf("%sN", i)
+    | PConstWasmF32(f) => Printf.sprintf("%sw", f)
+    | PConstWasmF64(f) => Printf.sprintf("%sW", f)
+    | PConstBool(true) => "true"
+    | PConstBool(false) => "false"
+    | PConstVoid => "void"
+    | PConstString(s) => Printf.sprintf("\"%s\"", String.escaped(s))
+    };
+  Doc.text(print_c);
+}
+and print_ident = (ident: Identifier.t) => {
+  switch (ident) {
+  | IdentName(name) =>
+    if (infixop(name) || prefixop(name)) {
+      Doc.concat([Doc.lparen, Doc.text(name), Doc.rparen]);
+    } else {
+      Doc.text(name);
+    }
+  | IdentExternal(externalIdent, second) =>
+    Doc.concat([
+      print_ident(externalIdent),
+      Doc.text("."),
+      Doc.text(second),
+    ])
+  };
+}
+
+and print_imported_ident = (ident: Identifier.t) => {
```

Comment:
Any reason why this should be separate from the regular ident printer?

Reply:
Nope, redundant and removed.